cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,221 @@
+  }
+
+  template <typename T, typename std::enable_if_t<!is_fixed_width<T>()>* = nullptr>
+  void CUDA_HOST_DEVICE_CALLABLE operator()(T const& key,
+                                            md5_intermediate_data* hash_state,
+                                            const md5_hash_constants_type* hash_constants,
+                                            const md5_shift_constants_type* shift_constants) const
+  {
+    CUDF_FAIL("Unsupported hash type");
+  }
+
+  void CUDA_HOST_DEVICE_CALLABLE operator()(Key const& key,
+                                            md5_intermediate_data* hash_state,
+                                            const md5_hash_constants_type* hash_constants,
+                                            const md5_shift_constants_type* shift_constants) const
+  {
+  }
+};
+
+/**
+ * @brief Specialization of MD5Hash operator for strings.
+ */
+template <>
+void CUDA_HOST_DEVICE_CALLABLE
+MD5Hash<cudf::string_view>::operator()(cudf::string_view const& key,
+                                       md5_intermediate_data* hash_state,
+                                       const md5_hash_constants_type* hash_constants,
+                                       const md5_shift_constants_type* shift_constants) const
+{
+  const uint32_t len  = (uint32_t)key.size_bytes();
+  const uint8_t* data = (const uint8_t*)key.data();
+
+  hash_state->message_length += len;
+
+  if (hash_state->buffer_length + len < 64) {
+    thrust::copy_n(thrust::seq, data, len, hash_state->buffer + hash_state->buffer_length);
+    hash_state->buffer_length += len;
+  } else {
+    uint32_t copylen = 64 - hash_state->buffer_length;
+    thrust::copy_n(thrust::seq, data, copylen, hash_state->buffer + hash_state->buffer_length);
+    md5_hash_step(hash_state, hash_constants, shift_constants);
+
+    while (len > 64 + copylen) {
+      thrust::copy_n(thrust::seq, data + copylen, 64, hash_state->buffer);
+      md5_hash_step(hash_state, hash_constants, shift_constants);
+      copylen += 64;
+    }
+
+    thrust::copy_n(thrust::seq, data + copylen, len - copylen, hash_state->buffer);
+    hash_state->buffer_length = len - copylen;
+  }
+}
+
+/**
+ * @brief Finalize MD5 hash including converstion to hex string.
+ */
+void CUDA_HOST_DEVICE_CALLABLE finalize_md5_hash(md5_intermediate_data* hash_state,
+                                                 char* result_location,
+                                                 const md5_hash_constants_type* hash_constants,
+                                                 const md5_shift_constants_type* shift_constants,
+                                                 const hex_to_char_mapping_type* hex_char_map)
+{
+  uint64_t full_length = (uint64_t)hash_state->message_length;
+  full_length          = full_length << 3;
+  thrust::fill_n(thrust::seq, hash_state->buffer + hash_state->buffer_length, 1, 0x80);
+
+  if (hash_state->buffer_length <= 55) {
```

Comment:
In the MD5 Hash the last value after the message is set to 1. Since we're only considering full byte input, that set bit requires a full byte + the full message length in bits requires 9 bytes.

Would `if (hash_state->buffer_length < 56) {` be more intuitive?

Reply:
I'd like for this to look something like this:
```
// 8 bytes for the message length plus a byte for the special character
constexpr int metadata_length = 9;
if (hash_state->buffer_length + metadata_length <= 64) {
```
(I probably did not get the comment or the name right)

Ideally 64 would also be named (I don't know what name that would be, but you probably do). Then, you can use these constants instead of of 64/55/56 in the code and it should become more readable (and less error-prone).